From 1c58d5b8e07344d3cf58de8ee8340b312fa9a372 Mon Sep 17 00:00:00 2001
From: Marc Kelly <mkelly@thegoodpenguin.co.uk>
Date: Tue, 24 Feb 2026 18:18:13 +0000
Subject: [PATCH 6/7] base: bootcache: Add kernel command line backend

Backend accepts data via the kernel commandline in the format of a
tag=value pairing.

Values prefixed with 0x are treated as arbitary long hex data.

eg.
bootcache_backend_cmdline.cache_data="teststring=TestValue,testbinary_long=0xABADCAFE000102030405A0A1A2A3A4A5,testbinary_short=0x01"

Signed-off-by: Marc Kelly <mkelly@thegoodpenguin.co.uk>
---
 .../admin-guide/kernel-parameters.txt         |   5 +
 drivers/base/Kconfig                          |   5 +
 drivers/base/Makefile                         |   1 +
 drivers/base/bootcache_backend_cmdline.c      | 176 ++++++++++++++++++
 4 files changed, 187 insertions(+)
 create mode 100644 drivers/base/bootcache_backend_cmdline.c

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index 62eb867caa58..42ec1f2d4ad3 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -658,6 +658,11 @@ Kernel parameters
     	bootcache.cache_disable= [BOOTCACHE]
         		Disable fetching of cached data.
 
+	bootcache_backend_cmdline.cache_data= [BOOTCACHE]
+			Supply cache data via commandline in the format of
+			comma seperated	<key>=<value> pairs.
+			Hex values prefixed with 0x are treated as arbitary raw data.
+
 	boot_delay=	[KNL,EARLY]
 			Milliseconds to delay each printk during boot.
 			Only works if CONFIG_BOOT_PRINTK_DELAY is enabled,
diff --git a/drivers/base/Kconfig b/drivers/base/Kconfig
index e0ed5a3a4131..8871ce4a39b9 100644
--- a/drivers/base/Kconfig
+++ b/drivers/base/Kconfig
@@ -101,6 +101,11 @@ config BOOTCACHE
 		  A backend that reads the cache data from reserved system memory.
 		  The reserved memory block is defined in the device tree and is
 		  assumed to be populated by the bootloader.
+ 
+	config BOOTCACHE_BACKEND_CMDLINE
+		bool "command line backend"
+		help
+		  A backend that reads the cache data from the kernel comamnd line.
 
 	endchoice
 	endif
diff --git a/drivers/base/Makefile b/drivers/base/Makefile
index d818e72df290..522adf2102ba 100644
--- a/drivers/base/Makefile
+++ b/drivers/base/Makefile
@@ -11,6 +11,7 @@ obj-$(CONFIG_AUXILIARY_BUS) += auxiliary.o
 obj-$(CONFIG_BOOTCACHE)    += bootcache.o
 obj-$(CONFIG_BOOTCACHE_BACKEND_TEST)    += bootcache_backend_test.o
 obj-$(CONFIG_BOOTCACHE_BACKEND_MEMORY)    += bootcache_backend_memory.o
+obj-$(CONFIG_BOOTCACHE_BACKEND_CMDLINE)    += bootcache_backend_cmdline.o
 obj-$(CONFIG_DEVTMPFS)	+= devtmpfs.o
 obj-y			+= power/
 obj-$(CONFIG_ISA_BUS_API)	+= isa.o
diff --git a/drivers/base/bootcache_backend_cmdline.c b/drivers/base/bootcache_backend_cmdline.c
new file mode 100644
index 000000000000..977b1c65515d
--- /dev/null
+++ b/drivers/base/bootcache_backend_cmdline.c
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: GPL-2.0
+#define DEBUG 1
+#include <linux/unaligned.h>
+#include <linux/crc32.h>
+#include <linux/init.h>
+#include <linux/io.h>
+#include <linux/kernel.h>
+#include <linux/slab.h>
+#include <linux/hashtable.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/of_device.h>
+#include <linux/of_reserved_mem.h>
+#include <linux/platform_device.h>
+#include <linux/string.h>
+#include <linux/types.h>
+#include <linux/errno.h>
+#include <linux/kconfig.h>
+#include <linux/kobject.h>
+#include <linux/sysfs.h>
+
+#include <linux/bootcache.h>
+
+#define DRIVER_NAME "bootcache_cmdline_backend"
+
+
+static char *cache_data = NULL;
+module_param(cache_data, charp, 0444);
+MODULE_PARM_DESC(cache_data, "Bootcache Data strings");
+
+/*
+ * This function processes the loaded data and adds each entry to the
+ * system cache via the callbck.
+ */
+static int cmdline_backend_load_cache(void)
+{
+	char *working_str, *pair;
+	int j, ret;
+
+	if (!cache_data) {
+	pr_info("%s: No data string provided.\n", DRIVER_NAME);
+		return 0;
+	}
+
+	pr_info("%s: Received data: %s\n", DRIVER_NAME, cache_data);
+
+	working_str = kstrdup(cache_data, GFP_KERNEL);
+	if (!working_str) {
+		return -ENOMEM;
+	}
+	while ((pair = strsep(&working_str, ",")) != NULL) {
+		char *name, *value;
+		struct bootcache_entry *new_entry = NULL;
+		size_t data_length, key_length, char_len;
+		u8 *data_buffer = NULL, *src = NULL, *dest = NULL;
+
+	if (!*pair)
+		continue;
+
+	/* Split into name and value */
+	name = strsep(&pair, "=");
+	value = pair;
+
+	if (name && value) {
+
+			key_length = strlen(name);
+
+		if (strstarts(value, "0x")) {
+		pr_info("%s: Found Hex data - %s = %s\n", DRIVER_NAME, name, value);
+				value +=2;
+				char_len = strlen(value);
+				if (char_len % 2 != 0) {
+				pr_err("%s: Key %s has invalid odd-length hex string\n", DRIVER_NAME, name);
+				continue;
+				}
+				data_length = char_len / 2;
+				data_buffer = (u8 *)kzalloc(data_length, GFP_KERNEL);
+				if (!data_buffer) {
+				return -ENOMEM;
+				}
+				if (hex2bin(data_buffer, value, data_length) != 0) {
+					pr_err("%s: hex2bin() failed for key %s\n", DRIVER_NAME, name);
+					kfree(data_buffer);
+					return -EINVAL;
+				}
+
+		} else {
+		pr_info("%s: Found String data - %s = %s\n", DRIVER_NAME, name, value);
+
+		data_length = strlen(value);
+				data_buffer = kzalloc(data_length, GFP_KERNEL);
+				if (!data_buffer) {
+				return -ENOMEM;
+				}
+				/*
+			 * Source data is potentially unaligned, so we copy it with the correct
+			* access functions
+			*/
+			src = value;
+			dest = data_buffer;
+			for (j = 0; j < data_length; j++)
+				*dest++ = get_unaligned(src++);
+			}
+
+			new_entry = kzalloc(sizeof(*new_entry), GFP_KERNEL);
+			if (!new_entry) {
+				kfree(data_buffer);
+				return -ENOMEM;
+			}
+
+			new_entry->key = kzalloc(key_length+1, GFP_KERNEL);
+			if (!new_entry->key) {
+				kfree(data_buffer);
+				kfree(new_entry);
+				return -ENOMEM;
+			}
+
+			new_entry->len = data_length;
+			new_entry->data = data_buffer;
+
+			src = name;
+			dest = new_entry->key;
+			for (j = 0; j < key_length; j++)
+				*dest++ = get_unaligned(src++);
+
+			*dest = '\0';
+
+			pr_debug("%s: Setting up new entry with key: %s, data length is %zu\n",
+				DRIVER_NAME, new_entry->key, new_entry->len);
+
+			/* call the framework provided function */
+			ret = bootcache_add_entry(new_entry);
+			if (ret) {
+				kfree(new_entry->key);
+				kfree(new_entry->data);
+				kfree(new_entry);
+				ret = 0;
+			}
+	}
+	}
+
+error:
+	kfree(working_str);
+
+	return ret;
+}
+
+static struct bootcache_info cache_info = {
+	.name = "cmdline",
+	.load_cache = cmdline_backend_load_cache,
+};
+
+static int __init bootcache_backend_init(void)
+{
+	int ret;
+
+	ret = bootcache_register_backend(&cache_info);
+
+	if (ret < 0) {
+		pr_err("%s: bootcache_register_backend() failed with error %d\n",
+			DRIVER_NAME, ret);
+		return ret;
+	}
+	pr_info("%s: Backend loaded\n", DRIVER_NAME);
+
+	return 0;
+}
+
+#if IS_BUILTIN(CONFIG_BOOTCACHE_BACKEND_CMDLINE)
+core_initcall(bootcache_backend_init);
+#else
+module_init(bootcache_backend_init);
+#endif
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("mkelly@thegoodpenguin.co.uk");
\ No newline at end of file
-- 
2.43.0

